### PR TITLE
[DI] Deprecate underscore-services in YamlFileLoader

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | no
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

As discussed when introducing `_defaults`

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php [hunk 1]
```diff
@@ -239,6 +239,9 @@ private function parseDefaults(array &$content, $file)
      */
     private function parseDefinition($id, $service, $file, array $defaults)
     {
+        if (preg_match('/^_[a-zA-Z0-9_]*$/', $id)) {
+            @trigger_error(sprintf('Service names that start with an underscore are deprecated since Symfony 3.3 and will be reserved in 4.0. Rename the "%s" service or define it in XML instead.', $id), E_USER_DEPRECATED);
```
Comment (javiereguiluz) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: I don't think it's a good idea that the allowed service ids depend on the config format. In the past, we've worked hard to make all formats work the same in spite of their differences (for example, by adding PHP constants support for Yaml files). If we want to deprecate something, we could deprecate just the `_defaults` string as service id (whatever the config format used).

Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: There is little comparison between supporting a feature in all formats (eg constants), and excluding a very specific narrow case for one specific loader which misses a few oops of expressiveness (eg underscore-service, which are not a "feature").
We also stated previously that we don't want parity with all loader formats - quite the contrary in fact: we want to embrace the full expressiveness of each format. Here, it's exactly a matter of expressiveness. Thus falls into this policy to me - thus only the yaml loader should be affected.

**Comment by jvasseur** [comment]
Maybe we could use a yaml tag for default values (or other non-service keys) instead and keep allowing anything for service names.

**Comment by nicolas-grekas** [comment]
We wondered about tags for eg `_defaults` in #21194, but that didn't look good. What would it look like to you?

**Comment by jvasseur** [comment]
Probably something like this
```yaml
services:
  defaults: !default
    public: false
  Foo\Bar: [@foo]
```
Any service tagged with default would be used as default instead of a service. And the key would be ignored.

**Comment by nicolas-grekas** [comment]
That's what I meant by:

> that didn't look good

I keep this opinion :)

**Comment by jvasseur** [comment]
Yeah I don't like it that much either, but using underscore-names to convey special meanings looks to much like a hack to me to like it either.

**Comment by javiereguiluz** [comment]
Is it too late to create a new config option called `services_defaults` ? It could convey its meaning better and it'd remove the need for hacks and for the YAML/XML unparity:

```yaml
services_defaults:
    public: false

services:
    Foo\Bar: [@foo]
```

**Comment by nicolas-grekas** [comment]
@javiereguiluz this discussion already happened, see #21071 :)

**Comment by stof** [comment]
@javiereguiluz this is *also* a BC break, as it makes this key forbidden for DI extensions (and it is much harder to write a BC layer, as we would be unable to detect this case as the DI extension could expose a config similar to our defaults)

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Pull request merged**